### PR TITLE
Improvement: Add get_cp_state method to iso15118 interface controller and include cp_status handler

### PR DESCRIPTION
**Pull request by ikaratass**
improvement on [AB#2162](https://dev.azure.com/switch-ev/09ab882d-ca4b-4b5b-b5b8-80603c12e5b2/_workitems/edit/2162)

### CODE REVIEW

File: iso15118/secc/states/iso15118_2_states.py [hunk 1]
```diff
@@ -1278,10 +1279,31 @@ async def process_message(
...
+            # Before closing the contactor, we need to check to
             # ensure the CP is in state C or D
+            cp_state = await self.comm_session.evse_controller.get_cp_state()
+            if cp_state not in [CpState.C2, CpState.D2]:
+                logger.info(
+                    f"Cp state is not C2 or D2, state is {cp_state} ."
+                    f"waiting for C2 or D2"
+                )
+                # wait for state C or D 250ms
+                await asyncio.sleep(0.25)
```
Comment (shalinnijel2) [inline iso15118/secc/states/iso15118_2_states.py:1294]: Just so I understand this correctly - is this 250ms from the specification? Or do we actually have a little less than 3 seconds to detect state C2 or D2?

Comment (ikaratass) [reply iso15118/secc/states/iso15118_2_states.py:1294]: this is from spec req number V2G2-847.  << The EV shall signal CP State C or D no later than 250ms after sending the first PowerDeliveryReq with ChargeProgress equals "Start" within V2G Communication SessionPowerDeliveryReq. >>

**Review by shalinnijel2 — CHANGES_REQUESTED** [review]

File: iso15118/secc/states/iso15118_2_states.py (hunk 1 above)
Comment (ikaratass) [reply iso15118/secc/states/iso15118_2_states.py:1290]: Done, actually, it was there, I  just rearranged it.

**Review by shalinnijel2 — DISMISSED** [review]

**Comment by shalinnijel2** [comment]
I missed to see this - would be good to add a reference to [V2G2-847] - "The EV shall signal CP State C or D no later than 250ms after sending the first PowerDeliveryReq with ChargeProgress equals "Start" within V2G Communication SessionPowerDeliveryReq." where we have the 250ms sleep.

File: tests/secc/states/test_iso15118_2_states.py [hunk 2]
```diff
@@ -334,3 +335,24 @@ async def test_power_delivery_set_hlc_charging(
         )
 
         self.comm_session.evse_controller.set_hlc_charging.assert_called_with(False)
+
+    @pytest.mark.parametrize(
+        "get_state_return_value, expected_next_state",
+        [
+            (CpState.C2, CurrentDemand),
```
Comment (shalinnijel2) [inline tests/secc/states/test_iso15118_2_states.py:342]: Would be good to check C1, D1 ends up in Terminate as well

File: iso15118/secc/states/iso15118_2_states.py [hunk 3]
```diff
@@ -1554,6 +1565,32 @@ async def process_message(
 
         self.expecting_power_delivery_req = False
 
+    async def wait_for_state_c_or_d(self) -> bool:
+        STATE_C_TIMEOUT = 0.25
```
Comment (shalinnijel2) [inline iso15118/secc/states/iso15118_2_states.py:1569]: Would be good to mention [V2G2-847] as the reason for 0.25.

**Review by shalinnijel2 — CHANGES_REQUESTED** [review]

File: iso15118/secc/states/iso15118_2_states.py (hunk 3 above)
Comment (ikaratass) [reply iso15118/secc/states/iso15118_2_states.py:1572]: Done

File: tests/secc/states/test_iso15118_2_states.py (hunk 2 above)
Comment (ikaratass) [reply tests/secc/states/test_iso15118_2_states.py]: done

**Review by shalinnijel2 — APPROVED** [review]

**Pull request merged**